### PR TITLE
mgocompat ObjectID, interface{}, slice, []byte, and symbol changes

### PR DESCRIPTION
**Pull request by iwysiu**
GODRIVER-1354
GODRIVER-1362
GODRIVER-1421
GODRIVER-1429
GODRIVER-1430

### CODE REVIEW

**Comment by iwysiu** [comment]
Sorry that this ended up being so long, I can break it up into smaller units if that's easier.
This also includes some things I missed in MapCodec

File: bson/bsoncodec/slice_codec.go [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+	case bsontype.Array:
+	case bsontype.Null:
+		val.Set(reflect.Zero(val.Type()))
+		return vr.ReadNull()
+	case bsontype.Type(0), bsontype.EmbeddedDocument:
+		if val.Type().Elem() != tE {
+			return fmt.Errorf("cannot decode document into %s", val.Type())
+		}
+	case bsontype.Binary:
+		if val.Type().Elem() != tByte {
```
Comment (divjotarora) [inline bson/bsoncodec/slice_codec.go]: This is saying this code should only work if the target slice type is `[]byte`. However, this should be handled by `ByteSliceCodec.DecodeValue`. Is this not the case?

File: bson/bsonoptions/empty_interface_codec_options.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+var defaultDecodeType = reflect.TypeOf(primitive.D{})
+
+// EmptyInterfaceCodecOptions represents all possible options for interface{} encoding and decoding.
+type EmptyInterfaceCodecOptions struct {
+	DecodeDefaultType  *reflect.Type // Specifies if the default type for decoding. Defaults to bson.D.
+	DecodeUnpackBinary *bool         // Specifies if Old and Generic type binarys should default to []slice instead of primitive.Binary. Defaults to false.
```
Comment (divjotarora) [inline bson/bsonoptions/empty_interface_codec_options.go]: New name suggestion: `DecodeBinaryAsByteSlice`? Or if that's too long, `DecodeBinaryAsSlice`?

File: bson/bsonrw/value_reader.go [hunk 3]
```diff
@@ -373,7 +373,7 @@ func (vr *valueReader) ReadBinary() (b []byte, btype byte, err error) {
 		return nil, 0, err
 	}
 
-	if btype == 0x02 {
+	if btype == 0x02 && length > 4 {
```
Comment (divjotarora) [inline bson/bsonrw/value_reader.go]: Can you add a comment for the rationale for this check?

File: bson/bsoncodec/empty_interface_codec.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	"go.mongodb.org/mongo-driver/bson/bsonrw"
+	"go.mongodb.org/mongo-driver/bson/bsontype"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+var defaultEmptyInterfaceCodec = NewEmptyInterfaceCodec()
+
+// EmptyInterfaceCodec is the Codec used for interface{} values.
+type EmptyInterfaceCodec struct {
+	DecodeDefaultType  reflect.Type
```
[the remark this answers is not in the document]
Comment (iwysiu) [reply bson/bsoncodec/empty_interface_codec.go]: done

**Pull request merged**